File: LRU Cache.py
class LRUCache:

    def __init__(self, capacity: int):
        self.dp = {}
        self.keys = {}
        self.capacity = capacity
        self.least_used = deque()
        self.rolling_key = 0

    def get(self, key: int) -> int:
        if key in self.dp:
            self.keys[key] = self.rolling_key
            self.least_used.append((self.rolling_key, key))
            self.rolling_key += 1
            return self.dp[key]
        return -1

    def put(self, key: int, value: int) -> None:
        if key in self.dp:
            self.dp[key] = value
            self.keys[key] = self.rolling_key
            self.least_used.append((self.rolling_key, key))
            self.rolling_key += 1
            return

        while len(self.dp) >= self.capacity:
            rolling_key, old_key = self.least_used.popleft()
            latest_rolling = self.keys[old_key]
            if latest_rolling != rolling_key:
                continue
            del self.dp[old_key]
            del self.keys[old_key]

        self.dp[key] = value
        self.keys[key] = self.rolling_key
        self.least_used.append((self.rolling_key, key))
        self.rolling_key += 1

# Your LRUCache object will be instantiated and called as such:
# obj = LRUCache(capacity)
# param_1 = obj.get(key)
# obj.put(key,value)

File: LRU Cache.cpp
class LRUCache {
public:
    int cap{0};
    int counter{0};
    std::unordered_map<int, int> mapping;
    std::unordered_map<int, int> freshness;
    std::vector<pair<int, int>> usages;

    LRUCache(int capacity) : cap(capacity) {
    }

    int get(int key) {
        if (mapping.contains(key)) {
            int fval = ++counter;
            usages.emplace_back(fval, key);
            push_heap(usages.begin(), usages.end(), std::greater<>{});
            freshness[key] = fval;
            return mapping[key];
        }
        return -1;
    }

    void put(int key, int value) {
        if (!mapping.contains(key)) {
            while (mapping.size() == cap) {
                pop_heap(usages.begin(), usages.end(), std::greater<>{});
                auto [fval, old_key] = std::move(usages.back());
                usages.pop_back();
                if (freshness[old_key] == fval) {
                    mapping.erase(old_key);
                }
            }
        }

        mapping[key] = value;
        int fval = ++counter;
        usages.emplace_back(fval, key);
        push_heap(usages.begin(), usages.end(), std::greater<>{});
        freshness[key] = fval;

    }
};

/**
 * Your LRUCache object will be instantiated and called as such:
 * LRUCache* obj = new LRUCache(capacity);
 * int param_1 = obj->get(key);
 * obj->put(key,value);
 */

